### PR TITLE
[Unix] Make UnixFileDescriptor copyable

### PR DESCRIPTION
**Pull request by zdobersek**
#### 75e50bc51ff72253bfc747349cc8414e0c76c1cd
<pre>
[Unix] Make UnixFileDescriptor copyable
<a href="https://bugs.webkit.org/show_bug.cgi?id=251358">https://bugs.webkit.org/show_bug.cgi?id=251358</a>

Reviewed by Michael Catanzaro.

Make UnixFileDescriptor copyable for now, matching similar types on other
platforms and enabling common behavior around classes that work with those
types.

The destructor is modified to invalidate any valid file descriptor value,
enabling correct procedure when a UnixFileDescriptor with an invalid fd value
is copied into a UnixFileDescriptor with a valid fd value.

It&apos;s still not great since copying these objects can lead to unnecessary
file descriptor churn, but that can be managed until there&apos;s some consensus
around how to operate with objects of these types (e.g. if move-only trait is
feasible across the board).

* Source/WTF/wtf/unix/UnixFileDescriptor.h:
(WTF::UnixFileDescriptor::UnixFileDescriptor):
(WTF::UnixFileDescriptor::operator=):

Canonical link: <a href="https://commits.webkit.org/259610@main">https://commits.webkit.org/259610@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/b24677427265c84eae349fa12e2a9c264bb23c69

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105280 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14365 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38173 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114541 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/174758 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15555 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5286 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97600 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/114465 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111036 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12038 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95008 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39508 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93904 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26642 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/81177 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/95101 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7696 "Built successfully") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28002 "Found 1 new test failure: media/video-playback-quality.html (failure)") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/93164 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/5425 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7791 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4581 "Found 3 new test failures: fast/forms/fieldset/fieldset-flexbox.html, fast/images/avif-as-image.html, fast/images/avif-heif-container-as-image.html (failure)") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30341 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13842 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47546 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/101866 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6628 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9579 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25420 "Passed tests") | 
| | | | |

### CODE REVIEW

**Review by philn — DISMISSED** [review]

**Comment by mcatanzaro** [comment]
I know this is needed to fix the build, but I unset the merge-queue label because I have some questions... sec.

File: Source/WTF/wtf/unix/UnixFileDescriptor.h [hunk 1]
```diff
@@ -65,6 +67,16 @@ class UnixFileDescriptor {
...
     }
 
+    UnixFileDescriptor& operator=(const UnixFileDescriptor& o)
+    {
+        if (&o == this)
+            return *this;
+
+        this->~UnixFileDescriptor();
+        new (this) UnixFileDescriptor(o);
+        return *this;
```
Comment (mcatanzaro) [inline Source/WTF/wtf/unix/UnixFileDescriptor.h:77]: So it looks like you intend to duplicate the fd when it is copied, and I agree that's the only sensible way to implement a copy. But I don't think this actually works? In particular, the destructor is going to close the fd without setting m_value to 0. Then the constructor is going to attempt to dup() an invalid fd. Right?

So it needs to be reordered a bit, although you can't exactly reorder the constructor to run before the destructor....

**Review by mcatanzaro — COMMENTED** [review]
My instincts suggest UnixFileDescriptor should be noncopyable. Forcing callers to manually call UnixFileDescriptor::duplicate seems very useful to avoid mistakes.

I know you're trying to fix things after https://commits.webkit.org/259550@main, but this feels fragile.

File: Source/WTF/wtf/unix/UnixFileDescriptor.h (hunk 1 above)
Comment (zdobersek) [reply Source/WTF/wtf/unix/UnixFileDescriptor.h:77]: Yes, good catch. The destructor should set the value to -1.

**Comment by zdobersek** [comment]
> My instincts suggest UnixFileDescriptor should be noncopyable. Forcing callers to manually call UnixFileDescriptor::duplicate seems very useful to avoid mistakes.
> 
> I know you're trying to fix things after https://commits.webkit.org/259550@main, but this feels fragile.

I would prefer it to be noncopyable, but it's easier for everyone to match `MachSendRight` and similar types in having it copyable.

The problem is mostly `UnixFileDescriptor` being aliased to `IPC::Connection::Handle`, which again was done to keep it simple. But an intermediate movable and copyable type can be (re)introduced that wraps around `UnixFileDescriptor`, and use it there.

**Comment by mcatanzaro** [comment]
> I would prefer it to be noncopyable, but it's easier for everyone to match MachSendRight and similar types in having it copyable.

Your call.

File: Source/WTF/wtf/unix/UnixFileDescriptor.h (hunk 1 above)
Comment (mcatanzaro) [reply Source/WTF/wtf/unix/UnixFileDescriptor.h:77]: Well, actually it was fine: I just got confused. It's `this.m_value` that gets closed. Then it should be set to -1 by the constructor. Then `o.m_value` will get duplicated, and you've already checked to ensure `this` and `o` are not the same. It should be fine even without the change to the destructor.

**Review by mcatanzaro — APPROVED** [review]

**Comment by mcatanzaro** [comment]
> <- bad reviewer

No, don't say that. ;)

**Comment by zdobersek** [comment]
> > I would prefer it to be noncopyable, but it's easier for everyone to match MachSendRight and similar types in having it copyable.
> 
> Your call.

I'll add it later, either in the WebKit's IPC namespace or alongside this class.

File: Source/WTF/wtf/unix/UnixFileDescriptor.h [hunk 2]
```diff
@@ -65,10 +67,20 @@ class UnixFileDescriptor {
...
+    }
+
     ~UnixFileDescriptor()
     {
         if (m_value >= 0)
-            closeWithRetry(m_value);
+            closeWithRetry(std::exchange(m_value, -1));
```
Comment (kkinnunen-apple) [inline Source/WTF/wtf/unix/UnixFileDescriptor.h:83]: FWIW, nothing can touch the `m_value` after destructor, so the `exchange` part is dead code.

**Review by kkinnunen-apple — COMMENTED** [review]
Thanks for fixing! Sorry for the warnings, it'd be good to catch them in EWS!

> It's still not great since copying these objects can lead to unnecessary
file descriptor churn, but that can be managed until there's some consensus
around how to operate with objects of these types (e.g. if move-only trait is
feasible across the board).

Copying is not "churn". If one needs a descriptor copied, then it must be copied.
That's why there is a "dup" operation on unix fd in the first place -- it's not a move-only type. Same with Win32 file handle and a Mach port..

**Comment by mcatanzaro** [comment]
> Thanks for fixing! Sorry for the warnings, it'd be good to catch them in EWS!
> 
> > It's still not great since copying these objects can lead to unnecessary
> > file descriptor churn, but that can be managed until there's some consensus
> > around how to operate with objects of these types (e.g. if move-only trait is
> > feasible across the board).
> 
> Copying is not "churn". If one needs a descriptor copied, then it must be copied. That's why there is a "dup" operation on unix fd in the first place -- it's not a move-only type. Same with Win32 file handle and a Mach port..

Well, semantically yes. But in practice, you really don't want to ever dup a file descriptor _unintentionally_, and that's not unlikely if it happens in the copy constructor or copy assignment operator. That's why having a separate function to create a copy seems safer here. But oh well. :)

(BTW, the EWS didn't catch this because the EWS use GCC, and this is a Clang warning. The EWS actually does detect GCC warnings now since https://commits.webkit.org/255961@main, which covers 98% of warnings, but of course it will miss any warning that does not occur with whatever particular version of GCC that the EWS runs.)

File: Source/WTF/wtf/unix/UnixFileDescriptor.h (hunk 2 above)
Comment (mcatanzaro) [reply Source/WTF/wtf/unix/UnixFileDescriptor.h:83]: Right, Zan added that because I tricked him into thinking it was necessary... my bad.

Comment (mcatanzaro) [reply Source/WTF/wtf/unix/UnixFileDescriptor.h:83]: I'm going to unset merge-queue one more time just to make sure Zan has seen this. Sorry. :P

**Comment by philn** [comment]
> > Thanks for fixing! Sorry for the warnings, it'd be good to catch them in EWS!
> 
> (BTW, the EWS didn't catch this because the EWS use GCC, and this is a Clang warning. The EWS actually does detect GCC warnings now since https://commits.webkit.org/255961@main, which covers 98% of warnings, but of course it will miss any warning that does not occur with whatever particular version of GCC that the EWS runs.)
> 

Some of the WPE/GTK EWS bots actually use clang, but there's no way to tell which and it's a kind-of lottery system currently.

It would be nice though to have a dedicated EWS check-box for WPE or GTK clang builds.

**Comment by philn** [comment]
This is becoming a bit off-topic, but I'd argue we should make clang default on linux EWS bots, because it's faster than GCC and it seems better at detecting issues (like this one).

**Comment by aperezdc** [comment]
> This is becoming a bit off-topic, but I'd argue we should make clang default on linux EWS bots, because it's faster than GCC and it seems more suited for detecting issues (like this one).

I very much agree with this.

**Comment by mcatanzaro** [comment]
> This is becoming a bit off-topic, but I'd argue we should make clang default on linux EWS bots, because it's faster than GCC and it seems more suited for detecting issues (like this one).

I'd rather not, because we'll be inundated with -Wreturn-type and -Wredundant-move warnings again. We already get clang coverage for cross-platform code via the Apple EWS, but the Linux EWS are the only coverage for GCC.

**Comment by dpino** [comment]
The reason why we use GCC on Linux is because it's the most commonly used compiler.

At some point we switched several post-commit build bots to Clang while keeping others with GCC (Debian Stable, Ubuntu). We learned later that some LayoutTests can produce different results depending on what compiler was used. See: https://bugs.webkit.org/show_bug.cgi?id=218220 (test passes if binary was built with Clang but it's a flakey with GCC) (cc @clopez).

After this experience we deployed two post-commit bots that build WebKitGTK/WPE using Clang:
  - [GTK-Linux-64-bit-Release-Clang-Build](https://build.webkit.org/#/builders/629)](https://build.webkit.org/#/builders/629)
  - [WPE-Linux-64-bit-Release-Clang-Build](https://build.webkit.org/#/builders/628)](https://build.webkit.org/#/builders/628)

Our goal is to deploy a post-commit bot that also runs Layout and API tests using the Clang builds.

Perhaps, if warnings were reported as fatal errors for Clang these post-commit bots would have detected the error, making it easier to narrow down the commit that introduced the regression.

About deploying an EWS Clang build bot for WebKitGTK and WPE we can put it in our agenda if you think that's useful, but I would discard switching the EWS build bots to one compiler or another for the reasons explained above.

File: Source/WTF/wtf/unix/UnixFileDescriptor.h (hunk 2 above)
Comment (zdobersek) [reply Source/WTF/wtf/unix/UnixFileDescriptor.h:83]: I think it's fine to set it back to -1, it's not costly.

**Comment by mcatanzaro** [comment]
> Perhaps, if warnings were reported as fatal errors for Clang these post-commit bots would have detected the error, making it easier to narrow down the commit that introduced the regression.

I don't recommend making warnings fatal on _most_ post-commit bots, because you will lose layout test results until the warning is fixed, and that makes tracking down regressions much more difficult. But if you just want to use it for a few particular bots, I think passing --fatal-warnings to build-webkit should override the --no-fatal-warnings that I added in https://commits.webkit.org/255954@main.

**Pull request merged**